### PR TITLE
Add `PromiseValue` type

### PR DESCRIPTION
**Pull request by resynth1943**
Boom.

Closes #42.

### CODE REVIEW

File: source/promise-value.d.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * async function setValue () {
+ *   data = await asyncFunction();
+ * }
+ *
+ * setValue();
+ *
+ * // Here's an example of using this with non-Promise types.
+ * function getNumber () {
+ *  return 2;
+ * }
```
Comment (fregante) [inline source/promise-value.d.ts]: I’m not sure if other types follow this, but, could you lint this piece of code with XO? The style doesn’t pass it

**Comment by resynth1943** [comment]
@fregante How about now?

**Comment by fregante** [comment]
Looks better. How about we drop the asyncFunction mention though? This would let the user focus on this type and nothing else.

**Comment by resynth1943** [comment]
Good idea. One sec.

**Comment by fregante** [comment]
I mean the whole thing, the example should not include `ReturnType<typeof asyncFunction>`. As discussed earlier, it seems that every example of PromiseValue is exactly that `PromiseValue<ReturnType<typeof asyncFunction>>`, suggesting that this type is useless on its own (as opposed to just having `AsyncReturnType<typeof asyncFunction>>`). Is it the case?

**Comment by resynth1943** [comment]
Thanks for your comment, @fregante. I'll say thanks for your contribution, however I'd still like to see this Promise-unpacking type implemented as a separate module to the variant that unpacks the return value from functions. Moreover, I still believe this type has uses on its own.

**Comment by fregante** [comment]
> Moreover, I still believe this type has uses on its own.

Right, I’m not disagreeing, I’m just asking to show a real use in the examples as well.

File: source/promise-value.d.ts [hunk 2]
```diff
@@ -4,21 +4,15 @@ If the type is not a `Promise`, the type itself is returned.
...
-
-// Here's an example of using this with non-Promise types.
-function getNumber(): number {
- return 2;
-}
-
-const number: PromiseValue<ReturnType<typeof getNumber>> = getNumber();
+// Here's an example that shows how this type reacts to non-Promise types.
+type SyncData = PromiseValue<string>;
+let syncData: SyncData = getSyncData();
```
Comment (resynth1943) [inline source/promise-value.d.ts]: As you can see, this example does not mention unpacking the return type of a function.

Comment (fregante) [reply source/promise-value.d.ts]: Yeah that works, thanks for making the change.

**Comment by sindresorhus** [comment]
Looks good 👍

**Comment by resynth1943** [comment]
Thanking you. How about we go for `AsyncReturnType` on top of this? ♥️

**Comment by sindresorhus** [comment]
👍

**Comment by resynth1943** [comment]
Let's do it.

**Pull request merged**